src/nanorc/tui.py

```diff
@@ -0,0 +1,589 @@
+
+    def compose(self) -> ComposeResult:
+        # yield TitleBox("Status {}")
+        yield StatusDisplay()
+
+class TreeDisplay(Static): pass
+
+class TreeView(Static):
+    rctree = reactive('none')
+    
+    def __init__(self, hostname, **kwargs):
+        super().__init__(**kwargs)
+        self.hostname = hostname
+        
+    def compose(self) -> ComposeResult:
+        yield TitleBox("Apps")
+        #yield Vertical(TreeDisplay(), id='verticaltree')
+        yield TreeDisplay()
+    
+    async def update_rctree(self) -> None:
+        async with httpx.AsyncClient() as client:
+            r = await client.get(f'{self.hostname}/nanorcrest/tree', auth=("fooUsr", "barPass"))
+        if r.json() == "I'm busy!":
+            return
+        #raise ValueError(r.json())
+        #Format is {'children': [...], 'name': 'foonode'} where the elements of children have the same structure
+        importer = DictImporter()
+        data = importer.import_(r.json())
+        the_text = Text("")
+        for pre, _, node in RenderTree(data):
+            the_line = Text("")
+            working = True
+            state_str = ''
+            style = ''
+            if node.errored:
+                state_str += "ERROR - "
+                style = 'bold red'
+                working = False
+
+            pathstring = ""
+            for n in node.path:     #The path is provided as a tuple of nodes, with the parents first
+                pathstring = pathstring + '.' + n.name
+
+            async with httpx.AsyncClient() as client:       #Checks the state of each node
+                fullpath = f'{self.hostname}/nanorcrest/node/' + pathstring
+                r = await client.get(fullpath, auth=("fooUsr", "barPass"))
+                node_data = r.json()
+
+                if 'process_state' in node_data:    #Only app nodes have this one
+                    if node_data['process_state'] != "alive":
+                        state_str += f"{node_data['process_state']} - "
+                        style = 'bold red'
+                        working = False 
+
+            state_str += f"{node.state}"
+            if not node.included:
+                state_str += " - excluded"
+                style = 'bright_black'
+                working = False
+            if node.state != "none" and working:        #A none state will have the default colour (white)
+                style = "green"
+            state_str += '\n'
+            the_text.append(f"{pre}{node.name}: ", style=(style))
+            the_text.append(state_str, style=(style))
+
+        self.rctree = the_text     #This is a string representation of the tree
+
+    def watch_rctree(self, rctree:str) -> None:
+        tree_display = self.query_one(TreeDisplay)
+        tree_display.update(rctree)
+
+    def on_mount(self) -> None:
+        self.set_interval(0.1, self.update_rctree)
+
+class Command(Static):
+    commands = reactive([])
+
+    class NewLog(Message):    
```

Comment:
Message classes need to be defined in the body of the class that produces them, so NewLog has to be defined inside of Command, since the command widget is what is sending the message.